src/ripple/basics/base_uint.h

```diff
@@ -314,7 +314,7 @@ class base_uint
         Hasher& h, base_uint const& a) noexcept
     {
         // Do not allow any endian transformations on this memory
-        h(a.pn, sizeof(a.pn));
+        h(a.pn.data (), sizeof(a.pn));
```

Comment:
Given:

```
std::array<T, N> a;
T b[n];```

are we guaranteed that `sizeof(a) == sizeof(b)`.

Reply:
I believe so. [According to CppRef](http://en.cppreference.com/w/cpp/container/array), "This container is an aggregate type with the same semantics as a struct holding a C-style array T[N] as its only non-static data member. "

Reply:
Yes, I'm pretty sure that's the case.  In the C++17 standard Section 26.3.7.1, *Class template array overview* paragraph 1 says in part:

> An instance of array<T, N> stores N elements of type T, so that size() == N is an invariant.

Then paragraph 2 says:

> An array is an aggregate (11.6.1) that can be list-initialized with up to N elements whose
> types are convertible to T.

Between those two guarantees I think we're good.

Reply:
No, especially when `N == 0`.  That being said, the following program compiles on all of our target platforms:
```
#include <array>
#include <cstdint>

template <class T, std::size_t N>
void
test()
{
    std::array<T, N> a;
    T b[N];
    static_assert(sizeof(a) == sizeof(b), "");
}

int
main()
{
    test<char, 1>();
    test<char, 2>();
    test<char, 3>();

    test<short, 1>();
    test<short, 2>();
    test<short, 3>();

    test<int, 1>();
    test<int, 2>();
    test<int, 3>();

    test<long long, 1>();
    test<long long, 2>();
    test<long long, 3>();
}
```

Reply:
It seems that in practice on all major vendors, this holds for all `N` other than `N == 0` like @HowardHinnant alluded to. See [this StackOverflow post](https://stackoverflow.com/questions/25173462/what-is-the-sizeof-stdarraychar-n) for some more standardese-explanation.

Reply:
Yipes!  I missed zero.  Good point.

Reply:
Yeah, I went down this path thinking about the "degenerate" case where `N == 0` and then wondered if there are other cases where this could happen.

Reply:
Do people have a preference of how to defend against this?  I am leaning towards just adding a `static_assert(WIDTH != 0, "Cannot have 0-sized base_uint");` to avoid these sort of surprises.

Great observation about the `N == 0` case @nbougalis.